core/src/Kokkos_HPX.hpp

```diff
@@ -215,120 +169,52 @@ class HPX {
-    std::unique_lock<hpx::spinlock> l(m_active_parallel_region_count_mutex);
-    if (--m_active_parallel_region_count == 0) {
-      l.unlock();
-      m_active_parallel_region_count_cond.notify_all();
-    };
+  void print_configuration(std::ostream &os, bool /*verbose*/ = false) const;
+  instance_data &impl_get_instance_data() const noexcept {
+    KOKKOS_ASSERT(m_instance_data.get());
+    return *m_instance_data.get();
   }
-
-  static void impl_increment_active_parallel_region_count() {
-    std::unique_lock<hpx::spinlock> l(m_active_parallel_region_count_mutex);
-    ++m_active_parallel_region_count;
+  uint32_t impl_instance_id() const noexcept {
+    return impl_get_instance_data().m_instance_id;
   }
-#endif
 
-  void fence(
-      const std::string &name =
-          "Kokkos::Experimental::HPX::fence: Unnamed Instance Fence") const {
-    Kokkos::Tools::Experimental::Impl::profile_fence_event<
-        Kokkos::Experimental::HPX>(
-        name,
-        Kokkos::Tools::Experimental::Impl::DirectFenceIDHandle{
-            impl_instance_id()},
-        [&]() {
-#if defined(KOKKOS_ENABLE_HPX_ASYNC_DISPATCH)
-          impl_get_future().wait();
-          // Reset the future to free variables that may have been captured in
-          // parallel regions.
-          impl_get_future() = hpx::make_ready_future<void>();
-#endif
-        });
+  static bool in_parallel(HPX const & = HPX()) noexcept {
+    // TODO: Very awkward to keep track of. What should this really return?
+    return false;
   }
 
-  static void impl_static_fence(const std::string &name) {
-    Kokkos::Tools::Experimental::Impl::profile_fence_event<
-        Kokkos::Experimental::HPX>(
-        name,
-        Kokkos::Tools::Experimental::SpecialSynchronizationCases::
-            GlobalDeviceSynchronization,
-        [&]() {
-#if defined(KOKKOS_ENABLE_HPX_ASYNC_DISPATCH)
-          std::unique_lock<hpx::spinlock> l(
-              m_active_parallel_region_count_mutex);
-          m_active_parallel_region_count_cond.wait(
-              l, [&]() { return m_active_parallel_region_count == 0; });
-          // Reset the future to free variables that may have been captured in
-          // parallel regions (however, we don't have access to futures from
-          // instances other than the default instances, they will only be
-          // released by fence).
-          HPX().impl_get_future() = hpx::make_ready_future<void>();
-#endif
-        });
-  }
+  static void impl_decrement_active_parallel_region_count();
+  static void impl_increment_active_parallel_region_count();
 
-  static hpx::execution::parallel_executor impl_get_executor() {
-    return hpx::execution::parallel_executor();
-  }
+  void impl_instance_fence_locked(const std::string &name) const;
+  void impl_instance_fence(const std::string &name =
+                               "Kokkos::Experimental::HPX::impl_instance_fence:"
+                               " Unnamed Instance Fence") const;
```

Comment:
Why the default argument here rather than on the `fence()` member function?
Also not sure we would want the `impl_` on the tooling side.

Reply:
Unsure. Refactorings on top refactorings. Your suggestions sound good.

Reply:
Changed in https://github.com/kokkos/kokkos/pull/5628/commits/2b5ab0d554271892babc427b80358d7e1adcd293.